### ISSUE TITLE
Decrease RAM usage

### ISSUE THREAD

**Issue by n-ce** [falzonv/discreet-launcher#90]
Decrease the Amount of RAM used to atleast 30MB.. current usage : 70MB

**Comment by xi7** [comment]
![Screenshot_2021-07-12-09-37-25](https://user-images.githubusercontent.com/67331785/125288746-dcd86980-e2f4-11eb-8a46-43710e79ea7b.png)

i think it can be less than 10mb, really load too much memory here and last 4.4.0 very slow

**Comment by falzonv** [comment]
Hello,

Thank you for your feedback!

Unfortunately, most of the RAM usage is taken by the icons of the applications :-/
It is to avoid loading them each time the user opens favorites or drawer, to display the list immediately.
Since I don't want to remove icons, I am not sure where to go from there...

Will let the ticket opened for some time in case there are some ideas.

Best regards.

**Comment by n-ce** [comment]
Maybe it can preload a set of colors inherited from the app icons instead of the icons, and then when app drawer is opened the actual icons get loaded in under 2 secs. The average user won't notice.

**Comment by falzonv** [comment]
Interesting idea, but it would only trade some memory usage for a lot of CPU consumption.

Currently, all the icons are preloaded when the list of apps is refreshed. This process can be quite CPU intensive if the user has lots of apps because all their icons have to be handled (even more process if an icon pack is selected).
Hopefully, it happens only after some specific events and, once it is done, there is very limited CPU consumption to navigate in the launcher because all icons are already preloaded in memory and reused each time.

If we reload all the icons each time the app drawer is opened, it means that this consuming process will be restarted each time the user will open the drawer. Because a launcher is supposed to be the starting point, it will happen a lot so, if we do this, soon after a new ticket will be opened by someone else with the title "Decrease CPU usage".
There is also a high risk that even more RAM will be consumed: if for some reason the user opens/closes the app drawer several times in a row, each last set of icons will be created in memory before the previous sets will have the time to be cleared by the Android garbage collector mechanism.

> The average user won't notice.

On this I do not agree, even it was under 500ms, most users would have the feeling that the launcher is lagging when loading the app list (and they would be right...).

**Comment by xi7** [comment]
#4 can be useful

**Comment by drewski-00h** [comment]
If you have 1GB ram, even 70mb of RAM usage is 7% of the RAM.  That seems pretty fair for any app.

**Comment by falzonv** [comment]
> @xi7: #4 can be useful

As I explained in conclusion of ticket #4, disabling the hardware acceleration was causing the launcher to become very slow and there was no difference in memory consumption (in fact it was even higher sometimes).
=> This is not something I will do as it would cause a regression.

> @drewski-00h: If you have 1GB ram, even 70mb of RAM usage is 7% of the RAM. That seems pretty fair for any app.

Totally agree, I didn't dare to say it.
It is even lower on the screenshot above in this ticket (21.07MB out of 526.53MB represents 4% of the RAM).

---

Reading again ticket #4, I also retrieved the following conclusion which is still valid:

> Considering that it includes all installed applications and their icons, the memory usage is not so high. Of course it is higher than an application not keeping these data in memory, but this is a design choice to make the launcher as quick as possible while staying quite beautiful and user-friendly by showing icons.
>
> To be honest, I am also using this application as my main launcher since 2019 (only published it on GitHub <s>a month</s> *five months* ago) and I have never faced any issue related to memory.

Maybe this can be re-discussed if someday a user faces an "Out Of Memory" error which closes the launcher.
But for now I will close this ticket and focus on others from the list.
